mixins/anchor-manager.js

```diff
@@ -0,0 +1,169 @@
+// manually trigger this behavior as well.)
+//
+// The 4-second default can be changed if the component defines
+// an instance variable called `ATTRACT_ATTENTION_TO_ANCHOR_DURATION`,
+// which should be set to an integer value in milliseconds.
+//
+// Note that this mixin will *not* directly affect the component's
+// rendering in any way. This means that it is not responsible for e.g.:
+//
+// * Setting any CSS classes, transitions, or animations.
+// * Setting the `id` attribute of any elements.
+// * Creating permalinks.
+
+var React = require('react');
+
+var DEFAULT_ATTRACT_DURATION = 4000;
+
+// AnchorManager is a largely private singleton class that maintains a
+// registry mapping anchor names to React components and listens for
+// location hash changes.
+//
+// When a registered component needs to be navigated to, its
+// `handleNavigateToAnchor` method is called; when it needs to
+// be navigated from, its `handleNavigateFromAnchor` method is called.
+var AnchorManager = function() {
+  this.anchorMap = {};
+  this.initialized = false;
+  this.currentAnchor = null;
+  this.handleHashChange = this.handleHashChange.bind(this);
+};
+
+AnchorManager.prototype = {
+  register: function(component) {
+    var anchorId = component.props.anchorId;
+    if (process.env.NODE_ENV !== 'production' &&
+        anchorId in this.anchorMap) {
+      console.warn('Anchor for #' + anchorId + ' already registered.');
+    }
+    this.anchorMap[anchorId] = component;
+    if (!this.initialized) {
+      this.initialize();
+      this.initialized = true;
+    }
+    this.handleHashChange();
+    return component;
+  },
+  unregister: function(component) {
+    var anchorId = component.props.anchorId;
+    if (process.env.NODE_ENV !== 'production') {
+      if (!(anchorId in this.anchorMap)) {
+        console.warn('Anchor for #' + anchorId + ' does not exist.');
+      } else if (this.anchorMap[anchorId] !== component) {
+        console.warn('Anchor mismatch for #' + anchorId + '.');
+      }
+    }
+    if (this.currentAnchor === component) {
+      this.currentAnchor = null;
+    }
+    delete this.anchorMap[anchorId];
+  },
+  initialize: function() {
+    window.addEventListener('hashchange', this.handleHashChange);
+  },
+  getHash: function() {
+    return window.location.hash.slice(1);
+  },
+  handleHashChange: function() {
+    var newAnchor = this.anchorMap[this.getHash()];
+
+    if (this.currentAnchor) {
+      if (newAnchor === this.currentAnchor) {
+        return;
+      } else {
+        this.currentAnchor.handleNavigateFromAnchor();
+        this.currentAnchor = null;
+      }
+    }
+    if (newAnchor) {
+      this.currentAnchor = newAnchor;
+      this.currentAnchor.handleNavigateToAnchor();
+    }
+  }
+};
+
+var defaultAnchorManager = new AnchorManager();
+
+var AnchorManagerMixin = {
+  propTypes: {
+    anchorId: React.PropTypes.string,
+    anchorManager: React.PropTypes.object
```

Comment:
The `anchorManager` prop is only used for testing. I considered making it a [React context](http://www.tildedave.com/2014/11/15/introduction-to-contexts-in-react-js.html) but those are such a pain to test and I kinda want to hold off on using contexts more until their API is stabilized and actually documented in React. We can always migrate to using a context for this later, too.